node/src/exchange/client.rs

```diff
@@ -56,63 +96,697 @@ impl ExchangeClientHandler {
             return;
         };
 
-        let req_id = req_resp.send_request(&peer, request);
-        self.reqs.insert(req_id, ReqInfo { amount, respond_to });
+        let req_id = req_resp.send_request(&peer, request.clone());
+        let state = State {
+            request,
+            respond_to,
+        };
 
-        trace!("Request initiated");
+        self.reqs.insert(req_id, state);
+    }
+
+    fn send_head_request(
+        &mut self,
+        req_resp: &mut R,
+        request: HeaderRequest,
+        respond_to: OneshotResultSender<Vec<ExtendedHeader>, P2pError>,
+    ) {
+        const MAX_PEERS: usize = 10;
+        const MIN_HEAD_RESPONSES: usize = 2;
+
+        let Some(peers) = self.peer_tracker.best_n_peers(MAX_PEERS) else {
+            respond_to.maybe_send_err(P2pError::NoPeers);
+            return;
+        };
+
+        let mut rxs = Vec::with_capacity(peers.len());
+
+        for peer in peers {
+            let (tx, rx) = oneshot::channel();
+
+            let req_id = req_resp.send_request(&peer, request.clone());
+            let state = State {
+                request: request.clone(),
+                respond_to: tx,
+            };
+
+            self.reqs.insert(req_id, state);
+            rxs.push(rx);
+        }
+
+        // Choose the best HEAD.
+        //
+        // Algorithm: https://github.com/celestiaorg/go-header/blob/e50090545cc7e049d2f965d2b5c773eaa4a2c0b2/p2p/exchange.go#L357-L381
+        spawn(async move {
+            let mut resps: Vec<_> = join_all(rxs)
+                .await
+                .into_iter()
+                .filter_map(|v| v.ok()?.ok()?.into_iter().next())
```

Comment:
this get's me thinking, shouldn't we there punish peers responding with multiple requests rather than ignoring it? I think in p2p the more defensive code the better. if you agree please also add a test

Reply:
At this point everything will be of length 1 because it is validated in `decode_and_verify_responses` method

Reply:
added a test just to ensure